### PR TITLE
[Notification] Fix unable to update multiple channels per type

### PR DESCRIPTION
**Pull request by qnkhuat**
Previously, we assumed that alerts could only have one destination for each channel type. This is because we only have 1 slack or 1 email connection.

With HTTP, we can have more than 1 HTTP channel. This PR fixes it so that we can enable multiple HTTP channels for an alert.
[Context](https://metaboat.slack.com/archives/C07318NQ5M2/p1724296690717229)

### CODE REVIEW

File: src/metabase/models/pulse.clj [hunk 1]
```diff
@@ -450,54 +450,49 @@
...
+  (let [existing-channels                 (t2/select :model/PulseChannel :pulse_id (u/the-id notification-or-id))
+        channel-type+id->existing-channel (update-vals (group-by (juxt :channel_type :channel_id) channels) first)
+        channels                          (map-indexed
+                                           (fn [idx channel]
+                                             (assoc channel
+                                                    :channel_type   (keyword (:channel_type channel))
+                                                    :schedule_type  (keyword (:schedule_type channel))
+                                                    :schedule_frame (keyword (:schedule_frame channel))
+                                                    :pulse_id       (u/the-id notification-or-id)
+                                                    ;; channel from FE might not have an id, so we need to find the id
```
Comment (piranha) [inline src/metabase/models/pulse.clj:472]: is this current behavior? I wonder if it makes sense to change that that and then this function will be much simpler?

**Review by crisptrutski — APPROVED** [review]
LGTM

**Comment by qnkhuat** [comment]
yea the build CI is failing upstream too, I'll fix it there

File: src/metabase/models/pulse.clj (hunk 1 above)
Comment (qnkhuat) [reply src/metabase/models/pulse.clj]: turns out we don't need it at all :) removed

**Review by npfitz — APPROVED** [review]

File: src/metabase/models/pulse.clj [hunk 2]
```diff
@@ -450,54 +450,45 @@
...
+    (doseq [channel to-update]
+      (assert (:id channel) "Cannot update a PulseChannel without an :id")
+      (pulse-channel/update-pulse-channel! channel))
     (binding [pulse-channel/*archive-parent-pulse-when-last-channel-is-deleted* false]
-      ;; for each of our possible channel types call our handler function
-      (doseq [[channel-type] pulse-channel/channel-types]
-        (handle-channel channel-type)))))
+      (doseq [channel to-delete]
+        (assert (:id channel) "Cannot delete a PulseChannel without an :id")
+        (t2/delete! PulseChannel :id (:id channel))))))
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/models/pulse.clj]: yes, N+1 is not nice!

**Pull request merged**